File: go/algorithm/heap/heap.go
/* Hint:
 * parent(i) = i>>1, index i divided by 2 (integer division),
 * left(i) = i<<1, index i multiplied by 2,
 * right(i) = (i<<1)+1, index i multiplied by 2 and added by 1.
 */

package heap

import "fmt"

type Heap []int

func NewHeap(nums []int) *Heap {
	h := make(Heap, 0)
	h.Create(nums)
	return &h
}

func Sort(nums []int) []int {
	h := NewHeap(nums)
	result := make([]int, len(nums))
	for i := len(nums) - 1; i >= 0; i-- {
		result[i], _ = h.ExtractMax()
	}
	return result
}

func (h *Heap) String() string {
	return fmt.Sprintf("%v", (*h)[1:])
}

func (h *Heap) Create(nums []int) {
	*h = []int{-1}
	*h = append(*h, nums...)

	for i := len(nums) / 2; i >= 1; i-- {
		h.siftDown(i)
	}
}

func (h *Heap) Insert(num int) {
	*h = append(*h, num)
	h.siftUp(len(*h) - 1)
}

func (h *Heap) ExtractMax() (int, error) {
	if len(*h) < 1 {
		return 0, fmt.Errorf("Empty Heap")
	}

	result := (*h)[1]
	(*h)[1] = (*h)[len(*h)-1]
	*h = (*h)[:len(*h)-1]
	h.siftDown(1)
	return result, nil
}

func (h *Heap) UpdateKey(idx, newValue int) {
	if idx < 1 || idx >= len(*h) {
		return
	}
	(*h)[idx] = newValue
	h.siftUp(idx)
	h.siftDown(idx)
}

func (h *Heap) Delete(idx int) {
	if idx < 1 || idx >= len(*h) {
		return
	}

	(*h)[idx] = (*h)[1] + 1
	h.siftUp(idx)
	_, _ = h.ExtractMax()
}

func (h *Heap) siftUp(idx int) {
	parent := idx >> 1
	for idx > 1 && (*h)[idx] > (*h)[parent] {
		(*h)[idx], (*h)[parent] = (*h)[parent], (*h)[idx]
		idx = parent
		parent = idx >> 1
	}
}

func (h *Heap) siftDown(idx int) {
	isLeftLarger := false
	isRightLarger := false

	left := idx << 1
	if left < len(*h) && (*h)[left] > (*h)[idx] {
		isLeftLarger = true
	}

	right := left + 1
	if right < len(*h) && (*h)[right] > (*h)[idx] {
		isRightLarger = true
	}

	if isLeftLarger && isRightLarger {
		if (*h)[right] > (*h)[left] {
			(*h)[right], (*h)[idx] = (*h)[idx], (*h)[right]
			h.siftDown(right)
		} else {
			(*h)[left], (*h)[idx] = (*h)[idx], (*h)[left]
			h.siftDown(left)
		}
	} else if isRightLarger {
		(*h)[right], (*h)[idx] = (*h)[idx], (*h)[right]
		h.siftDown(right)
	} else if isLeftLarger {
		(*h)[left], (*h)[idx] = (*h)[idx], (*h)[left]
		h.siftDown(left)
	}
}

File: ts/algorithm/heap/heap.ts
export class Heap {
  static sort(nums: Array<number>): Array<number> {
    let h = new Heap(nums)
    let result = new Array(nums.length)
    for (let i = nums.length - 1; i >= 0; i--)
      result[i] = h.extractMax()
    return result
  }

  _data: Array<number>

  constructor(nums: Array<number>) {
    this._data = [-1].concat(nums)
    for (let i = this._data.length >> 1; i >= 1; i--)
      this.siftDown(i)
  }


  get data(): Array<number> {
    return this._data.slice(1)
  }

  insert(num: number) {
    this._data.push(num)
    this.siftUp(this._data.length - 1)
  }

  extractMax(): number {
    if (this._data.length < 1)
      return Number.NEGATIVE_INFINITY

    let result = this._data[1]
    this._data[1] = this._data[this._data.length - 1]
    this._data.pop()
    this.siftDown(1)
    return result
  }

  updateKey(idx: number, value: number) {
    if ((idx < 1) || (idx >= this._data.length))
        return

    this._data[idx] = value
    this.siftUp(idx)
    this.siftDown(idx)
  }

  delete(idx: number) {
    if ((idx < 1) || (idx >= this._data.length))
        return

    this._data[idx] = Number.POSITIVE_INFINITY
    this.siftUp(idx)
    this.extractMax()
  }

  siftUp(idx: number) {
    let parent = idx >> 1
    while ((idx > 1) && (this._data[idx] > this._data[parent])) {
      this.swap(idx, parent)
      idx = parent
      parent = idx >> 1
    }
  }

  siftDown(idx: number) {
    let isLeftLarger = false
    let isRightLarger = false

    let left = idx << 1
    if ((left < this._data.length) &&
        (this._data[left] > this._data[idx]))
      isLeftLarger = true

    let right = left + 1
    if ((right < this._data.length) &&
        (this._data[right] > this._data[idx]))
      isRightLarger = true

    if (isLeftLarger && isRightLarger) {
      if (this._data[left] > this._data[right]) {
        this.swap(left, idx)
        this.siftDown(left)
      } else {
        this.swap(right, idx)
        this.siftDown(right)
      }
    } else if (isLeftLarger) {
      this.swap(left, idx)
      this.siftDown(left)
    } else if (isRightLarger) {
      this.swap(right, idx)
      this.siftDown(right)
    }
  }

  swap(i: number, j: number) {
    let tmp = this._data[i]
    this._data[i] = this._data[j]
    this._data[j] = tmp
  }
}
